src/mem/address_space.h

```diff
@@ -0,0 +1,206 @@
+#include "../pal/pal.h"
+#include "ds/address.h"
+#include "ds/flaglock.h"
+
+#include <array>
+#include <iostream>
+namespace snmalloc
+{
+  template<typename Pal>
+  class AddressSpaceManager : Pal
+  {
+    /**
+     * Implements a power of two allocator, where all blocks are aligned to the
+     * same power of two as their size. This is what snmalloc uses to get
+     * alignment of very large sizeclasses.
+     *
+     * It cannot unreserve memory, so this does not require the
+     * usual complexity of a buddy allocator.
+     */
+
+    // There are a maximum of two blocks for any size/align in a range.
+    // One before the point of maximum alignment, and one after.
+    // As we can add multiple ranges the second entry's block may contain
+    // a pointer to subsequent blocks.
+    std::array<std::array<void*, 2>, bits::BITS> ranges = {};
+
+    // This is infrequently used code, a spin lock simplifies the code
+    // considerably, and should never be on the fast path.
+    std::atomic_flag spin_lock = ATOMIC_FLAG_INIT;
+
+    inline void check_block(void* base, size_t align_bits)
+    {
+      SNMALLOC_ASSERT(
+        base == pointer_align_up(base, bits::one_at_bit(align_bits)));
+      // All blocks need to be bigger than a pointer.
+      SNMALLOC_ASSERT(bits::one_at_bit(align_bits) >= sizeof(void*));
+      UNUSED(base);
+      UNUSED(align_bits);
+    }
+
+    void add_block(size_t align_bits, void* base)
+    {
+      check_block(base, align_bits);
+      if (ranges[align_bits][0] == nullptr)
```

Comment:
Note that `operator[]` is not bounds checking.  Do we want an explicit assert here for the size or to use `.at()` instead?

Reply:
Doesn't `at` use exceptions?  I'll add an assert.